File: library/src/main/java/renetik/android/lang/CSDoLater.java
package renetik.android.lang;

import android.os.Handler;

import static renetik.android.lang.CSLang.error;
import static renetik.android.lang.CSLang.exception;

public class CSDoLater {

    private static Handler handler;
    private final Runnable _runnable;

    public CSDoLater(Runnable runnable, int delayMilliseconds) {
        _runnable = runnable;
        if (!handler.postDelayed(_runnable, delayMilliseconds))
            error(exception("Runnable not run"));
    }

    public CSDoLater(Runnable runnable) {
        _runnable = runnable;
        if (!handler.post(_runnable)) error(exception("Runnable not run"));
    }

    public static void initializeHandler() {
        handler = new Handler();
    }

    public void stop() {
        handler.removeCallbacks(_runnable);
    }

}

File: library/renetik-android/src/main/java/renetik/android/lang/CSDoLater.kt
package renetik.android.lang

import android.os.Handler
import renetik.java.extensions.exception
import renetik.android.lang.CSLog.logError


private lateinit var handler: Handler

fun initializeHandler() {
    handler = Handler()
}

fun doLater(delayMilliseconds: Int, function: () -> Unit): CSDoLater {
    return CSDoLater(function, delayMilliseconds)
}

fun doLater(function: () -> Unit) = CSDoLater(function)

class CSDoLater {

    private val function: () -> Unit

    constructor(function: () -> Unit, delayMilliseconds: Int) {
        this.function = function
        if (!handler.postDelayed(this.function, delayMilliseconds.toLong()))
            logError(exception("Runnable not run"))
    }

    constructor(function: () -> Unit) {
        this.function = function
        if (!handler.post(this.function)) logError(exception("Runnable not run"))
    }

    fun stop() = handler.removeCallbacks(function)

}
